src/scheduler/fifo.c

```diff
@@ -671,6 +672,11 @@ scheduling_cycle(int sd, char *jobid)
 	schdlog(PBSEVENT_DEBUG, PBS_EVENTCLASS_REQUEST, LOG_DEBUG,
 		"", "Starting Scheduling Cycle");
 
+	if (threads == NULL) {
+		if (init_multi_threading() != 0)
```

Comment:
I had debated about this myself. I wanted the core scheduling library, or libpbssched to have the multithreading infrastructure, not the program pbs_sched. The scheduling library can be used by other programs for their own purposes, so I wanted it to be contained inside it. I didn't see any performance lag caused by the creation and deletion of threads every cycle, it takes microseconds, so I decided to keep it inside the library.

Reply:
I'd thought about that as well, I didn't do it there because then I didn't know where I would gracefully kill the threads. Right now, it happens in end_cycle_tasks(), if we remove it from there, we'd have to put it in the pbs_sched program's main, and I wanted to avoid that.

So I'd rather we keep it the way it is since it doesn't hurt performance, unless you can think of a way to keep them inside libpbssched and start and end them cleanly.

Reply:
Well, if you add the creation of threads in schedinit() (or something similar) you can also add a schedexit() in libpbssched which can be called by pbs_sched when it exits.
I don't see how it is helping us by recreating threads every cycle. IMO, Thread pool of worker threads should persist throughout the life of the process. Also, putting code in end_cycle_tasks does not guarantee any graceful cleanup of threads anyways, so I do not think it would be any worse.

Reply:
hmm .. ok, i looked at how the pbs start/stop script works, and it sends TERM to pbs_sched, so ya, putting it in end_cycle_tasks() won't really help terminate it gracefully. I'll do what you suggested and create the threads in schedinit(), and add graceful termination inside SIGTERM handler instead.

Reply:
Actually what you had before is fine.  During a scheduling cycle, the scheduler blocks all signals.  If you keep what you have, you need to terminate the threads in the QUIT part of schedule as well.